src/main/java/ch/njol/skript/expressions/ExprItemsIn.java

```diff
@@ -40,109 +42,132 @@
+		if (inventories instanceof Variable && !inventories.isSingle() && parseResult.mark != 1)
 			Skript.warning("'items in {variable::*}' does not actually represent the items stored in the variable. Use either '{variable::*}' (e.g. 'loop {variable::*}') if the variable contains items, or 'items in inventories {variable::*}' if the variable contains inventories.");
 		return true;
 	}
-	
-	@SuppressWarnings("null")
+
 	@Override
+	@SuppressWarnings("null")
 	protected Slot[] get(final Event e) {
-		final ArrayList<Slot> r = new ArrayList<>();
-		for (final Inventory invi : invis.getArray(e)) {
+		ArrayList<Slot> r = new ArrayList<>();
+		ItemType[] types = this.types == null ? null : this.types.getArray(e);
+		for (Inventory invi : inventories.getArray(e)) {
 			for (int i = 0; i < invi.getSize(); i++) {
-				if (invi.getItem(i) != null)
+				if (isAllowedItem(types, invi.getItem(i)))
 					r.add(new InventorySlot(invi, i));
 			}
 		}
 		return r.toArray(new Slot[r.size()]);
 	}
-	
+
 	@Override
 	@Nullable
-	public Iterator<Slot> iterator(final Event e) {
-		final Iterator<? extends Inventory> is = invis.iterator(e);
+	public Iterator<Slot> iterator(Event e) {
+		Iterator<? extends Inventory> is = inventories.iterator(e);
+		ItemType[] types = this.types == null ? null : this.types.getArray(e);
 		if (is == null || !is.hasNext())
 			return null;
 		return new Iterator<Slot>() {
 			@SuppressWarnings("null")
 			Inventory current = is.next();
-			
+
 			int next = 0;
-			
-			@SuppressWarnings("null")
+
 			@Override
+			@SuppressWarnings("null")
 			public boolean hasNext() {
-				while (next < current.getSize() && current.getItem(next) == null)
+				while (next < current.getSize() && !isAllowedItem(types, current.getItem(next)))
 					next++;
 				while (next >= current.getSize() && is.hasNext()) {
 					current = is.next();
 					next = 0;
-					while (next < current.getSize() && current.getItem(next) == null)
+					while (next < current.getSize() && !isAllowedItem(types, current.getItem(next)))
 						next++;
 				}
 				return next < current.getSize();
 			}
-			
+
 			@Override
 			public Slot next() {
 				if (!hasNext())
 					throw new NoSuchElementException();
 				return new InventorySlot(current, next++);
 			}
-			
+
 			@Override
 			public void remove() {
 				throw new UnsupportedOperationException();
 			}
 		};
 	}
-	
+
 	@Override
 	public boolean isLoopOf(final String s) {
 		return s.equalsIgnoreCase("item");
 	}
-	
+
 	@Override
 	public String toString(final @Nullable Event e, final boolean debug) {
-		return "items in " + invis.toString(e, debug);
+		return "items in " + inventories.toString(e, debug);
 	}
-	
+
 	@Override
 	public boolean isSingle() {
 		return false;
 	}
-	
+
 	@Override
 	public Class<Slot> getReturnType() {
 		return Slot.class;
 	}
-	
+
+	private boolean isAllowedItem(@Nullable ItemType[] types, @Nullable ItemStack item) {
+		if (types == null)
+			return item != null;
+		else if (item == null)
```

Comment:
```suggestion
		if (item == null)
```
Since we're dealing with return statements